### ISSUE TITLE
get Segmentation fault when run runcpu.sh

### ISSUE THREAD

**Issue by MZLCC** [karpathy/nanochat#150]
part execute output as below. 

Overriding: depth = 4
Overriding: max_seq_len = 1024
Overriding: device_batch_size = 1
Overriding: total_batch_size = 1024
Overriding: eval_every = 50
Overriding: eval_tokens = 4096
Overriding: core_metric_every = 50
Overriding: core_metric_max_per_task = 12
Overriding: sample_every = 50
Overriding: num_iterations = 50
Autodetected device type: mps
2025-10-22 16:39:49,038 - nanochat.common - INFO - Distributed world size: 1
Vocab size: 65,536
num_layers: 4
model_dim: 256
num_heads: 2
num_kv_heads: 2
Tokens / micro-batch / rank: 1 x 1024 = 1,024
Tokens / micro-batch: 1,024
Total batch size 1,024 => gradient accumulation steps: 1
Number of parameters: 36,700,160
Estimated FLOPs per token: 1.321206e+08
Using user-provided number of iterations: 50
Total number of training tokens: 51,200
Tokens : Params ratio: 0.00
Total training FLOPs estimate: 6.764573e+12
Scaling the LR for the AdamW parameters ∝1/√(256/768) = 1.732051
dev/runcpu.sh: line 56: 31902 Segmentation fault: 11  python -m scripts.base_train --depth=4 --max_seq_len=1024 --device_batch_size=1 --total_batch_size=1024 --eval_every=50 --eval_tokens=4096 --core_metric_every=50 --core_metric_max_per_task=12 --sample_every=50 --num_iterations=50

**Comment by MZLCC** [comment]
os: mac
arch: Apple M3 Pro

**Comment by cthrall** [comment]
Same here with a M1 Max running Sequoia 15.6.1.

Stack trace starts like this:
```
0   libobjc.A.dylib               	       0x1826e3930 objc_retain + 16
1   AGXMetalG13X                  	       0x1372b56e0 AGX::BlitDispatchContext<AGX::G13::Encoders, AGX::G13::Classes, AGX::G13::ObjClasses>::bindComputeResources(IOGPUMetalResource const* const*, IOGPUMetalResource const* const*, bool) + 92
2   AGXMetalG13X                  	       0x1372b6530 AGX::BlitDispatchContext<AGX::G13::Encoders, AGX::G13::Classes, AGX::G13::ObjClasses>::blitCDMBuffer(IOGPUMetalResource const*, unsigned long long, IOGPUMetalResource const*, unsigned long long, unsigned long, void const*, unsigned long) + 324
3   AGXMetalG13X                  	       0x1372aeeec -[AGXG13XFamilyBlitContext copyFromBuffer:sourceOffset:toBuffer:destinationOffset:size:] + 64
3   AGXMetalG13X                  	       0x1372aeeec -[AGXG13XFamilyBlitContext copyFromBuffer:sourceOffset:toBuffer:destinationOffset:size:] + 64
4   libtorch_cpu.dylib            	       0x1520cdd74 invocation function for block in at::mps::MPSStream::copy(id<MTLBuffer>, objc_object<MTLBuffer>, unsigned long, unsigned long, unsigned long, unsigned long long, at::mps::SyncType) + 176
5   libdispatch.dylib             	       0x18293985c _dispatch_client_callout + 16
6   libdispatch.dylib             	       0x18292f7a8 _dispatch_lane_barrier_sync_invoke_and_complete + 56
7   libtorch_cpu.dylib            	       0x1520cdea0 at::mps::MPSStream::copy_and_sync(id<MTLBuffer>, objc_object<MTLBuffer>, unsigned long, unsigned long, unsigned long, bool, unsigned long long) + 104
8   libtorch_cpu.dylib            	       0x152118998 at::native::mps::mps_copy_(at::Tensor&, at::Tensor const&, bool) + 10616
9   libtorch_cpu.dylib            	       0x14cf1e180 at::native::copy_(at::Tensor&, at::Tensor const&, bool) + 2304
10  libtorch_cpu.dylib            	       0x14dce2c7c at::_ops::copy_::call(at::Tensor&, at::Tensor const&, bool) + 296
11  libtorch_cpu.dylib            	       0x14d25ace8 at::native::_to_copy(at::Tensor const&, std::__1::optional<c10::ScalarType>, std::__1::optional<c10::Layout>, std::__1::optional<c10::Device>, std::__1::optional<bool>, bool, std::__1::optional<c10::MemoryFormat>) + 964
12  libtorch_cpu.dylib            	       0x14d81d0a4 at::_ops::_to_copy::redispatch(c10::DispatchKeySet, at::Tensor const&, std::__1::optional<c10::ScalarType>, std::__1::optional<c10::Layout>, std::__1::optional<c10::Device>, std::__1::optional<bool>, bool, std::__1::optional<c10::MemoryFormat>) + 188
13  libtorch_cpu.dylib            	       0x14d81bfbc at::_ops::_to_copy::call(at::Tensor const&, std::__1::optional<c10::ScalarType>, std::__1::optional<c10::Layout>, std::__1::optional<c10::Device>, std::__1::optional<bool>, bool, std::__1::optional<c10::MemoryFormat>) + 336
14  libtorch_cpu.dylib            	       0x14d9d0110 at::_ops::to_device::call(at::Tensor const&, c10::Device, c10::ScalarType, bool, bool, std::__1::optional<c10::MemoryFormat>) + 324
15  libtorch_python.dylib         	       0x113016980 torch::utils::(anonymous namespace)::internal_new_from_data(c10::TensorOptions, c10::ScalarType, std::__1::optional<c10::Device>, _object*, bool, bool, bool, bool) + 4244
16  libtorch_python.dylib         	       0x11301cedc torch::utils::tensor_ctor(c10::DispatchKey, c10::ScalarType, torch::PythonArgs&) + 500
17  libtorch_python.dylib         	       0x112a75154 torch::autograd::THPVariable_tensor(_object*, _object*, _object*) + 328
```

Tried the nightly build of PyTorch 2.10 and got the same result.

**Comment by jwnsu** [comment]
Please check closed issue #143 for resolution. You can also try PR #142.

**Comment by cthrall** [comment]
Yup, I just noticed #142 was merged so I pulled and that resolved the issue for me. Thanks!

**Comment by svlandeg** [comment]
Thanks for reporting back, @cthrall! I'll go ahead and close this issue, assuming it's been resolved 🙏
